comment/api/serializers.py

```diff
@@ -99,31 +101,69 @@ def get_reactions(obj):
 
 
 class CommentCreateSerializer(BaseCommentSerializer):
-
     class Meta:
         model = Comment
-        fields = ('id', 'user', 'content', 'parent', 'posted', 'edited', 'reply_count', 'replies', 'urlhash')
+        fields = ('id', 'user', 'email', 'content', 'parent', 'posted', 'edited', 'reply_count', 'replies', 'urlhash')
 
     def __init__(self, *args, **kwargs):
         self.model_name = kwargs['context'].get('model_name')
         self.app_name = kwargs['context'].get('app_name')
         self.model_id = kwargs['context'].get('model_id')
         self.user = kwargs['context'].get('user')
         self.parent_id = kwargs['context'].get('parent_id')
+        if kwargs['context']['request'].user.is_authenticated or not settings.COMMENT_ALLOW_ANONYMOUS:
+            del self.fields['email']
+
         super().__init__(*args, **kwargs)
 
-    def get_parent_object(self):
-        if not self.parent_id or self.parent_id == '0':
-            return None
-        return Comment.objects.get(id=self.parent_id)
+    def validate_email(self, value):
+        request = self.context['request']
+        if request.user.is_authenticated:
+            value = request.user.email
+
+        elif settings.COMMENT_ALLOW_ANONYMOUS:
+            if (not value):
+                raise serializers.ValidationError(
+                    _('Email is required for posting anonymous comments.'),
+                    code='invalid'
+                    )
+
+        return value.strip().lower()
 
     def create(self, validated_data):
-        return Comment.objects.create(
-            user=self.user,
-            content_object=get_model_obj(self.app_name, self.model_name, self.model_id),
-            parent=self.get_parent_object(),
-            content=validated_data.get("content")
-        )
+        request = self.context['request']
+        user = get_user_for_comment(request)
+        content = self.validated_data.get('content')
+        email = self.validated_data.get('email')
+        parent_id = self.parent_id
+        time_posted = timezone.now()
+        parent_comment = Comment.objects.get_parent_comment(parent_id)
+        model_object = get_model_obj(self.app_name, self.model_name, self.model_id)
+
+        comment = Comment(
+            content_object=model_object,
+            content=content,
+            user=user,
+            parent=parent_comment,
+            email=email,
+            posted=time_posted
+            )
+        dict_comment = {
+            'user': user,
+            'content': content,
+            'email': email,
+            'posted': str(time_posted),
+            'app_name': self.app_name,
+            'model_name': self.model_name,
+            'model_id': self.model_id,
+            'parent': parent_id
+        }
+        if settings.COMMENT_ALLOW_ANONYMOUS and (not user):
+            response_msg = process_anonymous_commenting(request, comment, dict_comment, api=True)
+            self.context['msg'] = response_msg
```

Comment:
what should I return here. I can obviously override the `post` method in the view of `generics.CreateAPIVIew` but I thought there may be an easier way to do this. Any suggestion from you would be great.

Reply:
What about None?

Reply:
What about None?

Reply:
None is obviously being returned here but it creates an issue with the `generics.CreateAPIView` inherited in the `view` function. It expects a model instance to be returned from the `create` function otherwise it shows an error.

Reply:
So I think we should use a different view for creating anonymous comment. I belive this would be cleaner since we will separate the logic.

Reply:
it just came to my mind, what about returning the comment instance, without saving it to db?

Reply:
> it just came to my mind, what about returning the comment instance, without saving it to db?

we may do this alongside the message that an email has been sent bla bla...